pkg/relayer/proxy/sync.go

```diff
@@ -152,13 +162,10 @@ func (server *relayMinerHTTPServer) serveSyncRequest(
 		"service_id", serviceId,
 		"supplier_operator_address", meta.SupplierOperatorAddress,
 	).Add(1)
-	defer func() {
-		startTime := time.Now()
-		duration := time.Since(startTime).Seconds()
-
+	defer func(startTime time.Time, statusCode *uint) {
 		// Capture the relay request duration metric.
-		relayer.RelaysDurationSeconds.With("service_id", serviceId).Observe(duration)
-	}()
+		relayer.CaptureRelayDuration(serviceId, startTime, int(*statusCode))
+	}(startTime, &statusCode)
```

Comment:
I think you should not make the anonymous function have those variables passed. Otherwise it would capture their current values.

```suggestion
	defer func() {
		relayer.CaptureRelayDuration(serviceId, startTime, int(*statusCode))
	}()
```

See this example: https://go.dev/play/p/uAilspJrlBj

Reply:
Check: https://go.dev/play/p/f7v-KsVAKc1 I'm using references for the `statusCode`, so it will be a pointer to the variable, not the value indeed, so it will have the right value at the end.

About `startTime` its ok to be a copy because that one will not change and should be the time at the very begin of the handler.